File: 4128-total-waviness-of-numbers-in-range-ii/total-waviness-of-numbers-in-range-ii.py
class Solution:
    waves = []

    # Precompute all valid 3-digit wave patterns
    for i in range(1000):
        right = i % 10
        mid = (i // 10) % 10
        left = (i // 100) % 10

        if (mid > left and mid > right) or \
           (mid < left and mid < right):
            waves.append(i)

    def totalWaviness(self, num1: int, num2: int) -> int:
        return self.count(num2) - self.count(num1 - 1)

    def count(self, num: int) -> int:
        if num < 100:
            return 0

        total = 0

        for pattern in self.waves:
            total += self.count_pattern(num, pattern)

        return total

    def count_pattern(self, num: int, pattern: int) -> int:
        count = 0
        multiplier = 1

        # Handle leading zero cases
        pattern_type = 1 if pattern < 100 else 0

        while multiplier * 100 <= num:

            prefix = num // (multiplier * 1000)
            current = (num // multiplier) % 1000
            suffix = num % multiplier

            if current > pattern:
                ways = prefix - pattern_type + 1
            elif current == pattern:
                ways = max(0, prefix - pattern_type)
                count += suffix + 1
            else:
                ways = max(0, prefix - pattern_type)

            count += ways * multiplier
            multiplier *= 10

        return count

File: 4128-total-waviness-of-numbers-in-range-ii/total-waviness-of-numbers-in-range-ii.js
/**
 * @param {number} num1
 * @param {number} num2
 * @return {number}
 */
var totalWaviness = function(num1, num2) {

    const waves = [];

    // Precompute all valid 3-digit wave patterns
    for (let i = 0; i < 1000; i++) {
        const right = i % 10;
        const mid = Math.floor(i / 10) % 10;
        const left = Math.floor(i / 100);

        if ((mid > left && mid > right) ||
            (mid < left && mid < right)) {
            waves.push(i);
        }
    }

    function count(num) {
        if (num < 100) return 0;

        let total = 0;

        for (const pattern of waves) {
            total += countPattern(num, pattern);
        }

        return total;
    }

    function countPattern(num, pattern) {
        let count = 0;
        let multiplier = 1;

        // Handle leading zero cases
        const patternType = pattern < 100 ? 1 : 0;

        while (multiplier * 100 <= num) {

            const prefix = Math.floor(num / (multiplier * 1000));
            const current = Math.floor(num / multiplier) % 1000;
            const suffix = num % multiplier;

            let ways = 0;

            if (current > pattern) {
                ways = prefix - patternType + 1;
            } else if (current === pattern) {
                ways = Math.max(0, prefix - patternType);
                count += suffix + 1;
            } else {
                ways = Math.max(0, prefix - patternType);
            }

            count += ways * multiplier;
            multiplier *= 10;
        }

        return count;
    }

    return count(num2) - count(num1 - 1);
};